### PR TITLE
Issue #825: Add PreferMethodReferenceCheck

### PR DESCRIPTION
**Pull request by AkMo3**
Resolves Issue #825: New Check "Prefer method reference to lambda"

Files not added from original :
 1. [whitelist.words](https://github.com/strkkk/checkstyle/blob/83182af23e9012993a10218adcb46eadce72ef41/.ci/jsoref-spellchecker/whitelist.words)
 2. [XpathRegressionPreferMethodReferenceTest.java](https://github.com/strkkk/checkstyle/blob/83182af23e9012993a10218adcb46eadce72ef41/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionPreferMethodReferenceTest.java)
 3. [SuppressionXpathRegressionPreferMethodReference.java](https://github.com/strkkk/checkstyle/blob/83182af23e9012993a10218adcb46eadce72ef41/src/it/resources/org/checkstyle/suppressionxpathfilter/prefermethodreference/SuppressionXpathRegressionPreferMethodReference.java)
 4. [SuppressionXpathRegressionPreferMethodReferenceExpression.java](https://github.com/strkkk/checkstyle/blob/83182af23e9012993a10218adcb46eadce72ef41/src/it/resources/org/checkstyle/suppressionxpathfilter/prefermethodreference/SuppressionXpathRegressionPreferMethodReferenceExpression.java)
 5. [SuppressionXpathRegressionPreferMethodReferenceExpression.java](https://github.com/strkkk/checkstyle/blob/83182af23e9012993a10218adcb46eadce72ef41/src/it/resources/org/checkstyle/suppressionxpathfilter/prefermethodreference/SuppressionXpathRegressionPreferMethodReferenceExpression.java)
 6. [PackageObjectFactory.java](https://github.com/strkkk/checkstyle/blob/83182af23e9012993a10218adcb46eadce72ef41/src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java)
 7. message.properties of all language except English.
 8. [PreferMethodReferenceCheck.xml](https://github.com/strkkk/checkstyle/blob/83182af23e9012993a10218adcb46eadce72ef41/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/PreferMethodReferenceCheck.xml)
 9. [xdocs/checks.xml](https://github.com/strkkk/checkstyle/blob/83182af23e9012993a10218adcb46eadce72ef41/src/xdocs/checks.xml)
 10. [xdocs/config_coding.xml](https://github.com/strkkk/checkstyle/blob/83182af23e9012993a10218adcb46eadce72ef41/src/xdocs/config_coding.xml)
 11. [InputPreferMethodReferenceStatementsDefault.java](https://github.com/strkkk/checkstyle/blob/83182af23e9012993a10218adcb46eadce72ef41/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceStatementsDefault.java), suprisingly this file is not used anywhere.
 
 checkstyle/checkstyle PR link: https://github.com/checkstyle/checkstyle/pull/9785
 checkstyle/contribution PR link: https://github.com/checkstyle/contribution/pull/552

### CODE REVIEW

**Comment by AkMo3** [comment]
@romani I need to update [checks-sevntu-error.xml](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/checks-sevntu-error.xml) and [checkstyle_sevntu_checks.xml](https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_sevntu_checks.xml) as travis fetchs them directly. For that I need to make 2 seperate PR for each repo, wouldn't it be better if these files are imported here only, and then directly used?

**Comment by romani** [comment]
let me see ...

meanwhile please resolve problem:
https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/763553835#L753
```
[INFO] --- jacoco-maven-plugin:0.7.1.201405082137:check (default-check) @ sevntu-checks ---

[INFO] Analyzed bundle 'sevntu-checks' with 81 classes

[WARNING] Rule violated for class com.github.sevntu.checkstyle.checks.coding.PreferMethodReferenceCheck: lines covered ratio is 0.15, but expected minimum is 1.00

[WARNING] Rule violated for class com.github.sevntu.checkstyle.checks.coding.PreferMethodReferenceCheck: branches covered ratio is 0.00, but expected minimum is 1.00
```

one more:
https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/763553837#L554
```
INFO] --- xml-maven-plugin:1.0.2:check-format (default) @ checkstyle ---

[ERROR] /home/travis/build/sevntu-checkstyle/sevntu.checkstyle/.ci-temp/checkstyle/./config/checkstyle_sevntu_checks.xml:18,2: Insert 4 spaces. Expected 4 found 0 spaces before start element <module>
```
this looks like misconfig of xml-maven-plugin as it try to validete files in `.ci-temp` but it should not .

**Comment by romani** [comment]
Please rebase to resolve second issue

**Comment by romani** [comment]
>[INFO] There are 2 errors reported by Checkstyle

Please send PR to checkstyle repo to fix such violations.

**Comment by AkMo3** [comment]
> meanwhile please resolve problem:
> https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/763553835#L753
> 
> ```
> [INFO] --- jacoco-maven-plugin:0.7.1.201405082137:check (default-check) @ sevntu-checks ---
> 
> [INFO] Analyzed bundle 'sevntu-checks' with 81 classes
> 
> [WARNING] Rule violated for class com.github.sevntu.checkstyle.checks.coding
> .PreferMethodReferenceCheck: lines covered ratio is 0.15, but expected minimum is 1.00
> 
> [WARNING] Rule violated for class com.github.sevntu.checkstyle.checks.coding.
> PreferMethodReferenceCheck: branches covered ratio is 0.00, but expected minimum is 1.00
> ```

Sir, this is weird but, jacoco is just not running this test. The `visitToken` is showing no coverage, this means this is not being run even one? Can you please help on how to solve this.

**Comment by romani** [comment]
Please update code to throw exception as first statement in this method and run a build, it will clearly show if it executed or no

**Comment by AkMo3** [comment]
https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/765450238

```
[INFO] --- jacoco-maven-plugin:0.7.1.201405082137:check (default-check) @ sevntu-checks ---

[INFO] Analyzed bundle 'sevntu-checks' with 81 classes

[WARNING] Rule violated for class com.github.sevntu.checkstyle.checks.coding.PreferMethodReferenceCheck: lines covered ratio is 0.15, but expected minimum is 1.00

[WARNING] Rule violated for class com.github.sevntu.checkstyle.checks.coding.PreferMethodReferenceCheck: branches covered ratio is 0.00, but expected minimum is 1.00
```

The test is not being run, as otherwise there would have been an exception.

File: sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/PreferMethodReferenceCheckTest.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.github.sevntu.checkstyle.checks.coding;
+
+import static com.github.sevntu.checkstyle.checks.coding.PreferMethodReferenceCheck.MSG_METHOD_REF;
+import static com.google.common.truth.Truth.assertThat;
+
+import org.junit.jupiter.api.Test;
```
Comment (romani) [inline sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/PreferMethodReferenceCheckTest.java:25]: https://github.com/sevntu-checkstyle/sevntu.checkstyle/blob/e26546952caf1bec4688bb219562db15120b1dfc/sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/AvoidConstantAsFirstOperandInConditionCheckTest.java#L24

Try to use old junit

**Review by romani — APPROVED** [review]
Item:

File: sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/PreferMethodReferenceCheckTest.java (hunk 1 above)
Comment (AkMo3) [reply sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/PreferMethodReferenceCheckTest.java]: Thanks, it solved the issue. But why was this happening?

**Comment by romani** [comment]
Please send PR to main repo to fix https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/765496011#L689

File: sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/PreferMethodReferenceCheckTest.java (hunk 1 above)
Comment (romani) [reply sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/PreferMethodReferenceCheckTest.java]: Junit5 requiring few more config, sevntu use old junit, so methods with new annotations are recognized by old junit

**Comment by AkMo3** [comment]
> Please send PR to main repo to fix https://travis-ci.org/github/sevntu-checkstyle/sevntu.checkstyle/jobs/765496011#L689

Sir, already done in https://github.com/checkstyle/checkstyle/pull/9785, also have done PR in contribution repo https://github.com/checkstyle/contribution/pull/551

File: sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/PreferMethodReferenceCheckTest.java (hunk 1 above)
Comment (AkMo3) [reply sevntu-checks/src/test/java/com/github/sevntu/checkstyle/checks/coding/PreferMethodReferenceCheckTest.java]: Ok. Thanks for letting me know.

**Pull request merged**